### PR TITLE
Generalize shouldHide in subthoughts

### LINKED ISSUE #1181: Autofocus + pinChildren doesn’t hide grandchildren
C38931DF-6468-4605-B449-E7532C9A93F2

## Steps to Reproduce

``` 
- a
  - =pinChildren
    - true
  - b
    - c
      - d
  - e
    - f
      - g
``` 

Set the cursor on `c`

## Current Behavior

`g` is visible

## Expected Behavior

`g` should be hidden

## Notes

Consider if the following rule might cover the case and guide the implementation:

If a thought that is not an ancestor of the cursor is hidden, then its descendents should always be hidden.

### PR DESCRIPTION
**Pull request by shresthabijay**
fixes #1181

### CODE REVIEW

**Comment by shresthabijay** [comment]
@raineorshine I have made the changes. Please review it. Thanks!

**Comment by raineorshine** [comment]
Confirmed fixed, though there is a regression in the table view. The comment you deleted was a hint.

We currently don't have any test coverage for autofocus, so that would be a useful place to start. We're going to need it if we turn up other bugs, want to refactor anything, or want to add new autofocus functionality (such as customizable expansion depth).

What do you think about RTL tests (e.g. as used in [toggleSort](https://github.com/cybersemics/em/blob/dev/src/shortcuts/__tests__/toggleSort.ts)) with [isElementHIddenByAutoFocus](https://github.com/cybersemics/em/blob/dev/src/util/isElementHiddenByAutoFocus.ts)? Lighter weight than e2e but still provides DOM testing.

Here's the regression:

## Steps to Reproduce

```js
- a
  - =view
    - Table
  - b
    - c
  - d
    - e
```

Set the cursor on `a/b/c`

## Current Behavior

2a0d9ef

`a/b/e` is hidden.

## Expected Behavior

`a/b/e` should be visible.

When the cursor is on a table grandchild leaf (column 2), other grandchildren of the table should be visible.

**Review by raineorshine — CHANGES_REQUESTED** [review]
Confirmed fixed, though there is a regression in the table view. The comment you deleted was a hint.

We currently don't have any test coverage for autofocus, so that would be a useful place to start. We're going to need it if we turn up other bugs, want to refactor anything, or want to add new autofocus functionality (such as customizable expansion depth).

What do you think about RTL tests (e.g. as used in [toggleSort](https://github.com/cybersemics/em/blob/dev/src/shortcuts/__tests__/toggleSort.ts)) with [isElementHIddenByAutoFocus](https://github.com/cybersemics/em/blob/dev/src/util/isElementHiddenByAutoFocus.ts)? Lighter weight than e2e but still provides DOM testing.

Here's the regression:

## Steps to Reproduce

```js
- a
  - =view
    - Table
  - b
    - c
  - d
    - e
```

Set the cursor on `a/b/c`

## Current Behavior

2a0d9ef

`a/b/e` is hidden.

## Expected Behavior

`a/b/e` should be visible.

When the cursor is on a table grandchild leaf (column 2), other grandchildren of the table should be visible.

**Comment by shresthabijay** [comment]
> Confirmed fixed, though there is a regression in the table view. The comment you deleted was a hint.
> 
> We currently don't have any test coverage for autofocus, so that would be a useful place to start. We're going to need it if we turn up other bugs, want to refactor anything, or want to add new autofocus functionality (such as customizable expansion depth).
> 
> What do you think about RTL tests (e.g. as used in [toggleSort](https://github.com/cybersemics/em/blob/dev/src/shortcuts/__tests__/toggleSort.ts)) with [isElementHIddenByAutoFocus](https://github.com/cybersemics/em/blob/dev/src/util/isElementHiddenByAutoFocus.ts)? Lighter weight than e2e but still provides DOM testing.
> 
> Here's the regression:
> 
> ## Steps to Reproduce
> ```js
> - a
>   - =view
>     - Table
>   - b
>     - c
>   - d
>     - e
> ```
> 
> Set the cursor on `a/b/c`
> 
> ## Current Behavior
> [2a0d9ef](https://github.com/cybersemics/em/commit/2a0d9efcfc9c09bfffe0a29da9c1a382d6c13c5b)
> 
> `a/b/e` is hidden.
> 
> ## Expected Behavior
> `a/b/e` should be visible.
> 
> When the cursor is on a table grandchild leaf (column 2), other grandchildren of the table should be visible.

I thought I tested this case but it turns out I tested with second columns having children.

**Comment by shresthabijay** [comment]
> What do you think about RTL tests (e.g. as used in [toggleSort](https://github.com/cybersemics/em/blob/dev/src/shortcuts/__tests__/toggleSort.ts)) with [isElementHIddenByAutoFocus](https://github.com/cybersemics/em/blob/dev/src/util/isElementHiddenByAutoFocus.ts)? Lighter weight than e2e but still provides DOM testing.

Turns out JSDOM doesn't load any stylesheets and also does not changes styles on class toggles. So it may not be easy to get style properties using `getComputedStyle`. I tried it and it doesn't have any necessary styles like opacity and color. I think we should use e2e test for this.

**Comment by shresthabijay** [comment]
@raineorshine I have made the changes. Both the cases are working properly now. We should add tests for the visibility of thoughts as a separate issues.

File: src/components/Subthoughts.tsx [hunk 1]
```diff
@@ -77,6 +77,10 @@ if (!toggleContextViewShortcut) throw new Error('toggleContextView shortcut not
 
 const PAGINATION_SIZE = 100
 
+/** Check if the given path is a leaf. */
+const isLeaf = (state: State, path: Path) => !getAllChildren(state, pathToContext(path))
+  .some((child: Child) => !isFunction(child.value))
```
Comment (raineorshine) [inline src/components/Subthoughts.tsx:82]: I'm not sure how this logic was working before, but it doesn't take into account `showHiddenThoughts`. You can use the `hasChildren` predicate selector instead:

https://github.com/cybersemics/em/blob/5101be65c855fe7e1a18adeb4efd62183d12c6c5/src/selectors/getChildren.ts#L46-L52

**Review by raineorshine — CHANGES_REQUESTED** [review]
`a/b/c` should be visible and dimmed.

image

**Comment by raineorshine** [comment]
> Turns out JSDOM doesn't load any stylesheets and also does not changes styles on class toggles. So it may not be easy to get style properties using `getComputedStyle`. I tried it and it doesn't have any necessary styles like opacity and color. I think we should use e2e test for this.

Ah, that's good to know. We could test against `distance-from-cursor-x` with JSDOM though. Not sure if that's ideal given that `distance-from-cursor` itself is the wrong semantics and needs to be refactored according to a more unified set of autofocus rules.

**Comment by shresthabijay** [comment]
> We could test against `distance-from-cursor-x` with JSDOM though. Not sure if that's ideal given that `distance-from-cursor` itself is the wrong semantics and needs to be refactored according to a more unified set of autofocus rules.

Totally agreed. It takes a while to understand how the autofocus affects the visibility and position of the thoughts. The main problem is that the `distance-from-cursor-x` approach only takes depth into account. But some thoughts needs to hidden or dimmed because of their relative position to the cursor and also may due to meta attributes. We need to do calculations to find the edge cases and fix them anyways. We can write a selector that determines the visibility and position of the thought given the current state and its path. It will be easier to test too.

**Review by raineorshine — APPROVED** [review]
Thank you, this works great, and the tests are a good short term solution until we have better visibility logic and detection!

File: src/components/__tests__/Visibility.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
+import { ReactWrapper } from 'enzyme'
```
[the remark this answers is not in the document]
Comment (shresthabijay) [reply src/components/__tests__/Visibility.ts:1]: Sure. I will keep that in mind. Thanks!
